File: examples/01_hello_world/workflow.py
"""Shared workflow definition for 01_hello_world.

Defines the graph and ops — imported by demo.py, serve_python.py, serve_rust.py.
"""

from hush.core import END, PARENT, START, GraphOp
from hush.core.ops.transform.func_op import op


@op
def greet(name: str):
    """Tạo greeting từ tên."""
    return {"greeting": f"Xin chào, {name}!"}


@op
def greet_en(name: str):
    """Tạo greeting tiếng Anh."""
    return {"greeting": f"Hello, {name}!"}


@op(rust="to_upper")
def upper(text: str):
    """Chuyển thành uppercase."""
    return {"result": text.upper()}


@op
def step_a():
    return {"a_result": "Kết quả A"}


@op
def step_b():
    return {"b_result": "Kết quả B"}


@op(rust="merge_two")
def merge(a: str, b: str):
    return {"combined": f"{a} + {b}"}


def build_hello():
    """Single greeting workflow."""
    with GraphOp(name="hello-world") as graph:
        g = greet(name=PARENT["name"])
        START >> g >> END
    return graph


def build_chain():
    """Two ops chained: greet → uppercase."""
    with GraphOp(name="two-steps") as graph:
        g = greet_en(name=PARENT["name"], outputs={"*": PARENT})
        u = upper(text=g["greeting"])
        START >> g >> u >> END
    return graph


def build_parallel():
    """Two ops in parallel → merge."""
    with GraphOp(name="parallel") as graph:
        a = step_a()
        b = step_b()
        m = merge(a=a["a_result"], b=b["b_result"])
        START >> [a, b] >> m >> END
    return graph

File: examples/rust_ops/src/ex01_hello_world/workflow.rs
//! ex01_hello_world — greeting, uppercase, parallel merge ops.

use hush_serve::hush_op;
use serde_json::{json, Value};

/// greet: name -> greeting = "Xin chao, {name}!"
#[hush_op]
pub fn greet(inputs: &Value) -> Value {
    let name = inputs["name"].as_str().unwrap_or("World");
    json!({"greeting": format!("Xin chào, {}!", name)})
}

/// greet_en: name -> greeting = "Hello, {name}!"
#[hush_op]
pub fn greet_en(inputs: &Value) -> Value {
    let name = inputs["name"].as_str().unwrap_or("World");
    json!({"greeting": format!("Hello, {}!", name)})
}

/// upper: text -> result = text.upper()
#[hush_op]
pub fn upper(inputs: &Value) -> Value {
    let text = inputs["text"].as_str().unwrap_or("");
    json!({"result": text.to_uppercase()})
}

/// step_a: () -> {"a_result": "Ket qua A"}
#[hush_op]
pub fn step_a(_inputs: &Value) -> Value {
    json!({"a_result": "Kết quả A"})
}

/// step_b: () -> {"b_result": "Ket qua B"}
#[hush_op]
pub fn step_b(_inputs: &Value) -> Value {
    json!({"b_result": "Kết quả B"})
}

/// merge: a, b -> {"combined": "{a} + {b}"}
#[hush_op]
pub fn merge(inputs: &Value) -> Value {
    let a = inputs["a"].as_str().unwrap_or("");
    let b = inputs["b"].as_str().unwrap_or("");
    json!({"combined": format!("{} + {}", a, b)})
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_greet() {
        let result = greet(&json!({"name": "Hush"}));
        assert_eq!(result["greeting"], "Xin chào, Hush!");
    }

    #[test]
    fn test_greet_en() {
        let result = greet_en(&json!({"name": "Hush"}));
        assert_eq!(result["greeting"], "Hello, Hush!");
    }

    #[test]
    fn test_upper() {
        let result = upper(&json!({"text": "hello world"}));
        assert_eq!(result["result"], "HELLO WORLD");
    }

    #[test]
    fn test_step_a() {
        let result = step_a(&json!({}));
        assert_eq!(result["a_result"], "Kết quả A");
    }

    #[test]
    fn test_step_b() {
        let result = step_b(&json!({}));
        assert_eq!(result["b_result"], "Kết quả B");
    }

    #[test]
    fn test_merge() {
        let result = merge(&json!({"a": "Hello", "b": "World"}));
        assert_eq!(result["combined"], "Hello + World");
    }
}
